component is left in broken state after redeployment of the pod

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:**
macOS

**Output of `odo version`:**
▶ odo version
odo v1.2.5 (552aad7ea)

## How did you run odo exactly?
```
▶ odo create  java-maven
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile existence [39413ns]
 ✓  Checking devfile compatibility [84734ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [89449ns]
 ✓  Validating devfile component [247861ns]

Please use `odo push` command to create the component with source deployed

▶ odo push

Validation
 ✓  Validating the devfile [188307ns]

Creating Kubernetes resources for component java-maven
 ✓  Waiting for component to start [2m]

Applying URL changes
 ✓  URL http-8080: http://http-8080-java-maven-tkral-test.apps.testocp4x.psiodo.net/ created

Syncing to component java-maven
 ✓  Checking files for pushing [5ms]
 ✓  Syncing files to the component [25s]

Executing devfile commands for component java-maven
 ✓  Executing mvn-package command "mvn package", if not running [1m]
 ✓  Executing run command "java -jar target/*.jar", if not running [2s]

Pushing devfile component java-maven
 ✓  Changes successfully pushed to component

▶ curl http://http-8080-java-maven-tkral-test.apps.testocp4x.psiodo.net/
Greetings from Spring Boot!%

▶ odo storage create mystorage --path=/data --size=1Gi
 ✓  Added storage mystorage to java-maven

Please use `odo push` command to make the storage accessible to the component

▶ odo push

Validation
 ✓  Validating the devfile [124117ns]

Creating Kubernetes resources for component java-maven
 ✓  Waiting for component to start [4m]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component java-maven
 ✓  Checking file changes for pushing [3ms]
 ✓  Syncing files to the component [2s]

Executing devfile commands for component java-maven
 ✗  Executing mvn-package command "mvn package", if not running [3s]
 ✗  Failed to start component with name java-maven. Error: Failed to create the component: command execution failed: unable to execute the run command: unable to exec command [/bin/sh -c mvn package]:
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.089 s
[INFO] Finished at: 2020-08-26T09:42:09Z
[INFO] ------------------------------------------------------------------------
[ERROR] The goal you specified requires a project to execute but there is no POM in this directory (/projects). Please verify you invoked Maven from the correct directory. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MissingProjectException
: error while streaming command: command terminated with exit code 1

```

## Expected behavior
source code should be present in container and my application  should be running after `odo push`


## Any logs, error output, etc?

build command while running `odo push` fails because the directory where the application source is supposed to be is empty.

```
▶ oc exec java-maven-748556d67-kmk7x  -- ls /projects
?
devfile.yaml
```

--- a/pkg/sync/adapter.go
+++ b/pkg/sync/adapter.go
@@ -108,7 +108,16 @@ func (a Adapter) SyncFiles(syncParameters common.SyncParameters) (isPushRequired
 			}
 		}
 
-		// run the indexer and find the modified/added/deleted/renamed files
+		// If the pod changed, reset the index, which will cause the indexer to walk the directory
+		// tree and resync all local files.
+		if syncParameters.PodChanged {
+			err = util.DeleteIndexFile(pushParameters.Path)
+			if err != nil {
+				return false, errors.Wrap(err, "unable to reset the index file")
+			}
+		}
+
+		// Run the indexer and find the modified/added/deleted/renamed files
 		ret, err = util.RunIndexer(pushParameters.Path, absIgnoreRules)
 		s.End(true)
 

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -5,6 +5,8 @@ import (
 	"net/http"
 	"os"
 	"path/filepath"
+	"reflect"
+	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -572,6 +574,55 @@ var _ = Describe("odo devfile push command tests", func() {
 			output = helper.CmdShouldPass("odo", commands...)
 			Expect(output).To(ContainSubstring("BUILD SUCCESS"))
 
+			// 4) Acquire files from remote container, filtering out target/* and .*
+			podName := cliRunner.GetRunningPodNameByComponent(cmpName, namespace)
+			output = cliRunner.Exec(podName, namespace, "find", sourcePath)
+			remoteFiles := []string{}
+			outputArr := strings.Split(output, "\n")
+			for _, line := range outputArr {
+
+				if !strings.HasPrefix(line, sourcePath+"/") {
+					continue
+				}
+
+				newLine, err := filepath.Rel(sourcePath, line)
+				Expect(err).ToNot(HaveOccurred())
+
+				newLine = filepath.ToSlash(newLine)
+				if strings.HasPrefix(newLine, "target/") || newLine == "target" || strings.HasPrefix(newLine, ".") {
+					continue
+				}
+
+				remoteFiles = append(remoteFiles, newLine)
+			}
+
+			// 5) Acquire file from local context, filtering out .*
+			localFiles := []string{}
+			err := filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+
+				newPath := filepath.ToSlash(path)
+
+				if strings.HasPrefix(newPath, ".") {
+					return nil
+				}
+
+				localFiles = append(localFiles, newPath)
+				return nil
+			})
+			Expect(err).ToNot(HaveOccurred())
+
+			// 6) Sort and compare the local and remote files; they should match
+			sort.Strings(localFiles)
+			sort.Strings(remoteFiles)
+
+			equal := reflect.DeepEqual(localFiles, remoteFiles)
+			if !equal {
+				fmt.Fprintf(GinkgoWriter, "Files mismatched between local and remote. local: {%v} remote: {%v}", localFiles, remoteFiles)
+				Fail("Mismatching files between local and remote")
+			}
 		}
 
 		It("Should ensure that files are correctly synced on pod redeploy, with force push specified", func() {

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -5,7 +5,6 @@ import (
 	"net/http"
 	"os"
 	"path/filepath"
-	"reflect"
 	"sort"
 	"strconv"
 	"strings"
@@ -617,12 +616,7 @@ var _ = Describe("odo devfile push command tests", func() {
 			// 6) Sort and compare the local and remote files; they should match
 			sort.Strings(localFiles)
 			sort.Strings(remoteFiles)
-
-			equal := reflect.DeepEqual(localFiles, remoteFiles)
-			if !equal {
-				fmt.Fprintf(GinkgoWriter, "Files mismatched between local and remote. local: {%v} remote: {%v}", localFiles, remoteFiles)
-				Fail("Mismatching files between local and remote")
-			}
+			Expect(localFiles).To(Equal(remoteFiles))
 		}
 
 		It("Should ensure that files are correctly synced on pod redeploy, with force push specified", func() {

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -10,7 +10,6 @@ import (
 	"time"
 
 	"github.com/openshift/odo/pkg/util"
-
 	"github.com/openshift/odo/tests/helper"
 	"github.com/openshift/odo/tests/integration/devfile/utils"
 
@@ -544,6 +543,43 @@ var _ = Describe("odo devfile push command tests", func() {
 			session := helper.CmdRunner("odo", "push", "-v", "5", "-f", "--namespace", namespace)
 			helper.WaitForOutputToContain("Non-readable POM", 180, 10, session)
 
+		})
+
+	})
+
+	Context("Verify files are correctly synced", func() {
+
+		// Tests https://github.com/openshift/odo/issues/3838
+		ensureFilesSyncedTest := func(namespace string, shouldForcePush bool) {
+			helper.CmdShouldPass("odo", "create", "java-springboot", "--project", namespace, cmpName)
+			helper.CopyExample(filepath.Join("source", "devfiles", "springboot", "project"), context)
+
+			fmt.Fprintf(GinkgoWriter, "Testing with force push %v", shouldForcePush)
+
+			// 1) Push a standard spring boot project
+			output := helper.CmdShouldPass("odo", "push", "--namespace", namespace)
+			Expect(output).To(ContainSubstring("Changes successfully pushed to component"))
+
+			// 2) Update the devfile.yaml, causing push to redeploy the component
+			helper.ReplaceString("devfile.yaml", "memoryLimit: 768Mi", "memoryLimit: 769Mi")
+			commands := []string{"push", "-v", "4", "--namespace", namespace}
+			if shouldForcePush {
+				// Test both w/ and w/o '-f'
+				commands = append(commands, "-f")
+			}
+
+			// 3) Ensure the build passes, indicating that all files were correctly synced to the new pod
+			output = helper.CmdShouldPass("odo", commands...)
+			Expect(output).To(ContainSubstring("BUILD SUCCESS"))
+
+		}
+
+		It("Should ensure that files are correctly synced on pod redeploy, with force push specified", func() {
+			ensureFilesSyncedTest(namespace, true)
+		})
+
+		It("Should ensure that files are correctly synced on pod redeploy, without force push specified", func() {
+			ensureFilesSyncedTest(namespace, false)
 		})
 
 	})

